Core/Object Arts/Dolphin/MVP/Views/Common Controls/ListView.cls

```diff
@@ -1474,41 +1474,51 @@ moreParamsAt: n put: anObject
 	^self moreParams at: n put: anObject!
 
 newSelectionsFromEvent: event
-	"Private - Answer the <integer> selection that would occur if the <MouseEvent>,
+	"Private - Answer a collection of the <integer> selections that would occur if the <MouseEvent>,
 	event, was was passed to the control."
 
 	| itemClicked anyKeysDown newSelections markIndex markRange markIsSelected |
 	itemClicked := self itemFromPoint: event position.
+	"In single-select mode, the new selection is always the item clicked, or nothing if the click is outside all items."
 	self isMultiSelect ifFalse: [^itemClicked ifNil: [#()] ifNotNil: [:item | Array with: item]].
 	newSelections := self selectionsByIndex.
-	""
 	anyKeysDown := event isShiftDown or: [event isCtrlDown].
-	((itemClicked isNil and: [anyKeysDown])
-		or: [event button ~= #left and: [anyKeysDown or: [newSelections includes: itemClicked]]])
-			ifTrue: [^newSelections].
-	""
+	"If the click is outside all items, and ctrl or shift is down, nothing changes."
+	(itemClicked isNil and: [anyKeysDown]) ifTrue: [^newSelections].
+	"A right or middle click on an already-selected item also does not change the selection,
+	otherwise it would be very frustrating to open the context menu with multiple items selected.
+	The control also chooses to leave the selection alone if modifiers are held during a right/middle click."
+	(event button ~= #left and: [anyKeysDown or: [newSelections includes: itemClicked]])
+		ifTrue: [^newSelections].
+	"Having taken care of the above edge cases, if no keys are down, the behavior is equivalent to the single-select case."
 	anyKeysDown ifFalse: [^itemClicked ifNil: [#()] ifNotNil: [:item | Array with: item]].
-	"Only Ctrl"
+	"Only Ctrl is down, so we flip the selection state of the clicked item."
 	event isShiftDown ifFalse: 
 			[newSelections := newSelections asSortedCollection.
```

Comment:
No, I don't think `selectionsByIndex` should never return an unsorted list. The purpose of making a `SortedCollection` is so that, in the case where we are _adding_ a selection, the value we ultimately return is in sorted order. I admit I didn't think of the pathological behavior of quicksort in this case, but I agree that the combination of Dolphin's mitigation and the (usually) small size of the collection make it not that important. 

I have often wondered about implementing an optimized `remove:ifAbsent:` on SortedCollection...something to stick in the back of my/our mind(s), I guess.